### PR TITLE
Add Duration#to_rounded_s method

### PR DESCRIPTION
**Pull request by jdickey**
This (internally) creates a copy of a Duration instance that has been
truncated so that at most two unit values are included, *and* rounded
from the original value so that those values are influenced as needed by
any additional values in the original.

For example:

```ruby
    t1 = (3.hours).to_rounded_s # "3 hours"
    t2 = (3.hours 49.seconds).to_rounded_s # "3 hours, 49 seconds"
    t3 = (3.hours 59.minutes 31.seconds)
    t3.to_rounded_s # "4 hours"
    t3.to_s # "3 hours, 59 minutes, 31 seconds"
```

The actual value of the Duration instance on which this method is called
is *not* affected.

### CODE REVIEW

**Comment by jdickey** [comment]
@kylewlacy, this code is **not yet ready** to merge; just reproduced and isolated a bug that demonstrates once again why you should never use simple types for anything non-trivial, _especially_ hashes that can have differing keys for the same conceptual value (such as, oh, say, `:hour` and `:hours`).

Should have a new commit within 24 hours from now.

**Comment by jdickey** [comment]
Here's the other shoe dropping. Say you've got a `Time` instance which is the result of `Time.since((1.day 4.hours 31.minutes).ago)`; call it `source` You want the units value `Hash` for that, so you call `source.to_units(*UNIT_ALIASES.keys)`, which hands you back the Hash

```ruby
{
  :millennium => 0,
  :millennia => 0,
  :century => 0,
  :centuries => 0,
  :decades => 0,
  :decade => 0,
  :year => 0,
  :years => 0,
  :month => 0,
  :months => 0,
  :week => 0,
  :weeks => 0,
  :days => 1,
  :day => 0,
  :hour => 4,
  :hours => 0,
  :minutes => 32,
  :minute => 0,
  :second => 20,
  :seconds => 0
}
```

You say "hey, there's way too much junque there", so you instead call `source.to_units(*UNIT_ALIASES.keys).keep_if { |_, v| v.nonzero? }`, which gives you

```ruby
{:days=>1, :hour=>4, :minutes=>32, :second=>20}
```

See anything interesting there? (edited)

**Comment by jdickey** [comment]
Right; that last commit solves the units-confusion bug that led me to write the [**not ready** comment](#issuecomment-384690936) above. The `#to_rounded_s` method should now Work Just Fine™.

I still stand by the ["other shoe dropping" comment](#issuecomment-384711439) earlier; that `#to_units` behaviour is *odd.*

**Comment by kylewlacy** [comment]
@jdickey Thanks for the PR! I haven't fully reviewed and tested it yet and I may not be able to for a couple of days/weeks, but I am interested in reviewing and hopefully merging your changes ASAP!

One question though: why introduce `#to_rounded_s` as a separate method instead of adding an option to `#to_s`? I'm not saying one approach is better, just curious to hear your thoughts on it, or if that was a necessary design decision for your use case, etc.

re the `#to_units` issue:  I agree that the current behavior is weird and should probably be filed as a bug (but you could probably argue that calling `#to_units` with equivalent, aliased units is currently unspecified.. and yes, using a sort of "unstructured" data type like a hashmap with arbitrary symbol keys doesn't help things). I imagine that falling back to the first occurrence of a unit alias would be more intuitive however.

**Comment by jdickey** [comment]
Hi, @kylewlacy, thanks for the positive response!

I added a new method, rather than modify the existing `#to_s` method for several reasons:

1. To minimise changes (and thus opportunity for typos/new bugs) to existing, working code;
2. To reinforce in the user's mind that this is conceptually different than a simple string conversion; we're changing the semantics from "give me a string representation of this Duration" to "give me a string representation of *what this Duration would be, approximated to two significant units*"; and
3. To minimise cognitive load for maintenance; i.e., to make it easier, if and when bugs are found in the new code, to reason about them and fix them with minimal risk of getting distracted/overwhelmed by keeping the rest of the code in your head at the same time. Having new classes in separate files helps with that, and (in my experience) largely the same argument can be made for adding methods to an existing class in a language like Ruby that supports reopening of classes.

I like to keep methods as small and SRP-compliant as practical, minimising the number of execution paths through a given method (and thus the number of tests needed to fully exercise the code). I've found in my experience that that saves a lot of ambiguity, uncertainty, and therefore opportunity for bugs to creep in. Keep the code as simple as possible to get the job done, but no simpler.

**Comment by jdickey** [comment]
Sure, I'll do that. I also just discovered an issue with the method when dealing with local time/UTC mismatches that cross a day span.

Consider the spec:

```ruby
  describe 'when the attribute is' do
    let(:actual) { article.date_as_str_for(:updated_at) }
    let(:article) do
      Fabricate :article, author_id: author.id,
                          created_at: 5.years.ago, # correct stamp checked?
                          updated_at: @offset
    end

    describe 'set to a value that is' do
      describe 'one month or more but less than one year' do
        it 'correctly reports rounded months and days' do
          @offset = 1.month.ago - 23.hours - 31.minutes
          expect(actual).must_equal '1 month, 1 day ago'
        end
      end # describe 'one month or more but less than one year'
    end # describe 'set to a value that is'
  end # describe 'when the attribute is'
```

I'm in Singapore, which is UTC+8, and the timestamps saved in the database are UTC (aka GMT). Depending on what the current local time is when I run that spec, it'll either report back '1 month, 1 day' or '1 month, 2 days' (as it just did when running at 23:51:40 SGT). It would be easy to argue that "well, we ought to just use one of the usual timestamp-munging ways for tests to never run into that problem"; the result being given is still _reasonably_ close, and it's not like it's ever going to say "1 month, 2 days" and then _later_ say "1 month, 1 day".

tl;dr Is this a bug in the test or a bug in the code, where fixing the latter would require knowledge of timezones which I'd argue is outside the current scope?

Assuming we're good with just faking the test, I'll have an updated `Duration` to you within 24 hours or so from the time I post this.

Thanks for your help.

**Comment by kylewlacy** [comment]
I think the example you posted is actually expected behavior due to the lossy nature of a duration. First, some background:

The core design principle behind `Duration` is that all durations can be exactly expressed in terms of 2 units: "seconds" and "months". An hour is 3,600 seconds, a week is 604,800 seconds, year is 12 months. But, a month isn't a specific number of seconds-- it changes based on the month! So, `Duration` literally has `@seconds` and `@months` instance variables to express this.

To make results more intuitive, most methods on `Duration` call `Duration#normalize`, which basically just approximates `@months` to `@seconds` (it assumes 30 days in a month by default). This gives us a lot of flexibility while maintaining as much precision as possible.

So, with that said, I believe your issue is with the `1.month.ago` part. Let me try to boil your problem down:

```ruby
expect(Time.since(1.month.ago - 23.hours - 31.minutes).to_rounded_s).to eq("1 month, 1 day")
expect(Time.since(1.month.ago).to_rounded_s).to eq("1 month")
expect(Time.since(1.month.before(Time.now)).to_rounded_s).to eq("1 month")
expect(Time.between(Time.now, Time.now - 1.month).to_rounded_s).to eq("1 month")
expect(Time.between(Time.now, Time.now - 1.month).to eq(1.month)
```

The last test case above demonstrates the same problem your test is seeing: going from `Duration` to `Time` and back is _not_ lossless. That's because `Time.between` (used by `#ago` and `#before`) just creates a `Duration` with the number of _seconds_ between each time. And, as stated above, since 1 month isn't exactly equal to a specific number of seconds, your test is dependent on the number of days in the month!

If I were you, I would either use fixed timestamps (`Time.before(time)`, etc.) or [Timecop](https://github.com/travisjeffery/timecop) for your specs in this case

---

Basically, all that is to say that I don't believe the snippet you posted is an issue with this PR in its current state (and I believe it is a fundamental issue with the concept of durations rather than a fixable bug... although I'm definitely open to suggestions if you think of an elegant solution for the library as a whole!)
